libnmstate/nm/profile.py

```diff
@@ -21,14 +21,948 @@
+
+    def is_master_iface(self):
+        return self.iface_state in MASTER_IFACE_TYPES
+
+    def is_slave_iface(self):
+        return self.iface_state.get(MASTER_METADATA)
+
+    @property
+    def context(self):
+        return self._ctx
+
+    def import_by_device(self, nmdev=None):
+        ac = get_device_active_connection(nmdev or self.nmdevice)
+        if ac:
+            if nmdev:
+                self.nmdevice = nmdev
+            self.profile = ac.props.connection
+            self._profile = ac.props.connection
+
+    def import_by_id(self, con_id=None):
+        if con_id:
+            self.con_id = con_id
+        if self.con_id:
+            self.profile = self._ctx.client.get_connection_by_id(self.con_id)
+
+    def import_by_uuid(self, uuid=None):
+        if uuid:
+            self.uuid = uuid
+        if self.uuid:
+            self.profile = self._ctx.client.get_connection_by_uuid(self.uuid)
+
+    def update(self, save_to_disk=True):
+        flags = NM.SettingsUpdate2Flags.BLOCK_AUTOCONNECT
+        if save_to_disk:
+            flags |= NM.SettingsUpdate2Flags.TO_DISK
+        else:
+            flags |= NM.SettingsUpdate2Flags.IN_MEMORY
+        action = f"Update profile: {self.profile.get_id()}"
+        user_data = action
+        args = None
+
+        self._ctx.register_async(action, fast=True)
+        self.profile.update2(
+            self.profile.to_dbus(NM.ConnectionSerializationFlags.ALL),
+            flags,
+            args,
+            self._ctx.cancellable,
+            self._update2_callback,
+            user_data,
+        )
+
+    def add(self, save_to_disk=True):
+        nm_add_conn2_flags = NM.SettingsAddConnection2Flags
+        flags = nm_add_conn2_flags.BLOCK_AUTOCONNECT
+        if save_to_disk:
+            flags |= nm_add_conn2_flags.TO_DISK
+        else:
+            flags |= nm_add_conn2_flags.IN_MEMORY
+
+        action = f"Add profile: {self.profile.get_id()}"
+        self._ctx.register_async(action, fast=True)
+
+        user_data = action
+        args = None
+        ignore_out_result = False  # Don't fall back to old AddConnection()
+        self._ctx.client.add_connection2(
+            self.profile.to_dbus(NM.ConnectionSerializationFlags.ALL),
+            flags,
+            args,
+            ignore_out_result,
+            self._ctx.cancellable,
+            self._add_connection2_callback,
+            user_data,
+        )
+
+    def delete(self):
+        if self._profile:
+            action = (
+                f"Delete profile: id:{self.profile.get_id()}, "
+                f"uuid:{self.profile.get_uuid()}"
+            )
+            user_data = action
+            self._ctx.register_async(action, fast=True)
+            self.profile.delete_async(
+                self._ctx.cancellable,
+                self._delete_connection_callback,
+                user_data,
+            )
+
+    def activate(self):
+        if self.uuid:
+            self.import_by_uuid()
```

Comment:
You should differetiate `NM.RemoteConnection` and `NM.SimpleConnection` like `self._remote_conn`  and `self._simple_conn`.

We only do import when `self._remote_conn` is undefined, not just import when uuid is defined.

You have to made it clear when we import:
 * Newly created profile.
 * Activation against `NM.Device` only like (checkpoint rollback).